### ISSUE TITLE
Casting input parameter creates compilation error

### ISSUE THREAD

**Issue by nenadvicentic** [microsoft/TypeScript#32972]
**TypeScript Version:**  3.5.3

**Search Terms:**
casting type assertion input parameter

**Code**

```ts
interface Evalable {
  eval(evalFunction: (node: Element) => string): string;
}
const page: Evalable = {
  eval: (evalFunction => evalFunction(new Element()))
};

// Error: Property 'innerText' does not exist on 'HTMLElement'.
const data = page.eval(<HTMLElement>(node: HTMLElement) => node.innerText);
```

**Expected behavior:**
Since `node` instide of the method is casted to `HTMLElement`, statement `node.innerText` should be valid.
**Actual behavior:**
Error "Property 'innerText' does not exist on 'HTMLElement'." is reported, even though `HTMLElement` has `innerText` property.

**Playground Link:** 
http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgKIDc4Bs4CMsoDeAsAFDLISZYAUV2AYgK4gJjAD2IAXMjSBwAmEXqgIBbCOACUyALwA+ZAGcwUUAHNpvVepAaA3GQC+ZBF1XIADnA0i01PAXnIS5StV51qzVuy7ySvRYvmycIPwQAO5oElJgNNJJJkakZAD06WhQUBxQvAAKuVbQYACeyADkoCDQACoQAB5glciCHBDKyAJglI3AlgGVABJ1ALIAMmIQkuCVAHRmFr2CcGBwLjZ288E0ADyjk9OzYAr8QvaHU3Eygd0X8zX1TWDSqUA

**Comment by dragomirtitian** [comment]
In this case `<HTMLElement>` is not a type assertion is is a type parameter list. This means that `<HTMLElement>(a: HTMLElement) => a.innerText)` is actually a generic function with an unconstrained type parameter `HTMLElement`, so `a` typed as this type parameter and will not have the property `innerText`.

If you want to cast the function to `HTMLElement` you need an extra set of `(<HTMLElement>((node: HTMLElement) => node.innerText))` (although it is not clear to me why you would do that)

**Comment by nenadvicentic** [comment]
@dragomirtitian I am not sure I follow you. Can you elaborate on that?

In my example IDE shows `a` to be of `HTMLElement` type and shows the error 

> Property 'innerText' does not exist on 'HTMLElement'.

At least error message and IDE intelisence is wrong.
I also tried your suggestion:
``` typescript
const data = page.eval(<HTMLElement>((node: HTMLElement) => node.innerText));
```
It just throws different type of error.

http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgKIDc4Bs4CMsoDeAsAFDLISZYAUV2AYgK4gJjAD2IAXMjXLxBMAtrmgBKZAF4AfMgDOYKKADm43ouUgVAbjIBfMgi6LkABzgqIvDNjwFpyEuUrVedas1bsu0ufSwvNk4QGgBGcXEDPVJjEFMAEzgwOEcLKwA6AJoAHgAJABUAWQAZVAJhCHAZGhoQDgTrZELS8ohK8ElZZHrGjNAQaAKIAA8wSJigA

**Comment by nmain** [comment]
> @dragomirtitian I am not sure I follow you. Can you elaborate on that?

Your original example is equivalent to this: [playground](http://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgKIDc4Bs4CMsoDeAsAFDLISZYAUV2AYgK4gJjAD2IAXMjSBwAmEXqgIBbCOACUyALwA+ZAGcwUUAHNpvVepAaA3GQC+ZBF1XIADnA0i01PAXnIS5StV51qzVuy7ySvRYvmycIPwQAO5oElJgNNJJJkakZDAsYQEwADwAKgr8QvZ5sm4UUBBgTFAgyALCAHSgINB5EAAeYKmmaaTmIJaCcGBwLjZ2jcE0MNKpQA).  Hopefully the different syntax and change in type alias makes it more clear what's going on.

> Intention in the code was this:
> 
> ```ts
> const data = page.eval(node => (node as HTMLElement).innerText);
> ```
> 
Then just do that.  Or, if you want to use the other type assertion syntax:

```ts
const data = page.eval(node => (<HTMLElement>node).innerText);
```

**Comment by nenadvicentic** [comment]
@nmain Ah, it's clear now from playground example. It's "anonymous generic function", so to speak.
